### PR TITLE
Reject unsupported migration

### LINKED ISSUE #15123: Reject process instance migration if any unsupported element is active
In the `ProcessInstanceMigration` command processor, reject the command if:
- an element is active that has an unsupported element type in the source process
- an element is active that has an unsupported element type in the target process (after mapping)

To start, all element types are unsupported.

Rejecting a command includes both appending a rejection with the rejection writer, as well as responding the rejection with the response writer.

Some tests must showcase this behavior.

Blocked by:
- https://github.com/camunda/zeebe/issues/15111

### PR DESCRIPTION
**Pull request by korthout**
## Description

Rejects the migration command when the process instance has an active element that is currently unsupported.

At this time, only `PROCESS` and `SERVICE_TASK` elements are supported. Active elements of any other type result in a rejected migration.

> [!NOTE]
> This pull request's large number of lines is due to a large set of example test cases. Most have a similar structure, so I looked into extracting it, but it didn't help the readability. I hope the examples can be easily read through.

> [!NOTE]
> While working on this, I wanted to reject all unsupported scenarios rather than just all unsupported elements. So I also tried to work on #15405 and #15406. However, those issues contain a lot of hidden complexity. For now, I've scoped these out of this pull request's related issue. We may want to consider implementing #15407 instead of the extracted issues.

## Related issues

closes #15123

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceModificationModifyProcessor.java [hunk 1]
```diff
@@ -125,16 +125,15 @@ public final class ProcessInstanceModificationModifyProcessor
...
           BpmnElementType.UNSPECIFIED,
           BpmnElementType.START_EVENT,
           BpmnElementType.SEQUENCE_FLOW,
           BpmnElementType.BOUNDARY_EVENT);
-  private static final Set<BpmnElementType> SUPPORTED_ELEMENT_TYPES =
-      Arrays.stream(BpmnElementType.values())
-          .filter(elementType -> !UNSUPPORTED_ELEMENT_TYPES.contains(elementType))
-          .collect(Collectors.toSet());
+  private static final EnumSet<BpmnElementType> SUPPORTED_ELEMENT_TYPES =
+      EnumSet.complementOf(UNSUPPORTED_ELEMENT_TYPES);
```
Comment (berkaycanbc) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceModificationModifyProcessor.java:135]: Nice one! 🚀

File: engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceMigrationMigrateProcessor.java [hunk 2]
```diff
@@ -100,7 +100,7 @@ public void processRecord(final TypedRecord<ProcessInstanceMigrationRecord> comm
     final var elementInstances = new ArrayDeque<>(List.of(processInstance));
     while (!elementInstances.isEmpty()) {
       final var elementInstance = elementInstances.poll();
-      migrateElementInstance(elementInstance, processDefinition, mappedElementIds);
+      tryMigrateElementInstance(elementInstance, processDefinition, mappedElementIds);
```
Comment (berkaycanbc) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceMigrationMigrateProcessor.java:103]: Really good one! 🚀

File: engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceMigrationMigrateProcessor.java [hunk 3]
```diff
@@ -103,6 +112,21 @@ public void processRecord(final TypedRecord<ProcessInstanceMigrationRecord> comm
         processInstanceKey, ProcessInstanceMigrationIntent.MIGRATED, value, command);
   }
 
+  @Override
+  public ProcessingError tryHandleError(
+      final TypedRecord<ProcessInstanceMigrationRecord> command, final Throwable error) {
+    if (error instanceof final UnsupportedElementMigrationException e) {
+      final String reason =
+          "Expected to migrate process instance '%s' but it contains an active element that is unsupported: %s"
```
Comment (berkaycanbc) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceMigrationMigrateProcessor.java:120]: ❌  The syntax is a bit blurry with indicating the element id. Would be better if we make the element id more explicit by saying:

`"Expected to migrate process instance '%s' but it contains an active element that is unsupported. Element id: %s"`

or something similar.

**Review by berkaycanbc — CHANGES_REQUESTED** [review]
Nice work @korthout! 🚀 

1️⃣ I asked for one change only.

👍 I agree with using exceptions. It conforms better with how our code handles expected errors. Also, I expect users not to send migration commands over an over again since we will present them a meaningful message why migration failed.

❓I didn't see a test case for `EVENT_BASED_GATEWAY`. Is it intentional? I think, a process might be waiting on active event based gateway, so would be good to test against.

❓What about the case where process definitions look like:

```java
 Bpmn.createExecutableProcess(SOURCE_PROCESS)
                     .startEvent()
                     .userTask("A")
                     .endEvent()
                     .done()
Bpmn.createExecutableProcess(TARGET_PROCESS)
                     .startEvent()
                     .subProcess(
                         "sub",
                         s ->
                             s.embeddedSubProcess()
                                 .startEvent()
                                 .serviceTask("B")
                                 .endEvent())
                     .endEvent()
                     .done()
```

the case above should be rejected as well right? Because, since sub process will not be migrated, the target process flow will not continue as expected. Will it be handled in another issue?

**Comment by korthout** [comment]
@berkaycanbc I've added the requested test case. As mentioned, the error message is adjusted in https://github.com/camunda/zeebe/pull/15417.

Please have another look.

**Review by berkaycanbc — APPROVED** [review]
LGTM @korthout! 🚀 

I also checked the error message reformatting in the other PR. 👍

**Comment by korthout** [comment]
bors merge

**Pull request merged**